change font size for code block in H1

yakupafsin opened the pull request:
@s77rt 

### Details
Created a new function to check if there is an ancestor node with the name 'h1' and applied on CodeRenderer.js to change code block font-size if it is inside h1 tag.

### Fixed Issues
$ https://github.com/Expensify/App/issues/30203
PROPOSAL: https://github.com/Expensify/App/issues/30203#issuecomment-1776167351


### Tests

1. Open the app
2. Open any report
3. Send a normal message and observe its font size
4. Send a heading message (add # in beginning)
5. Verify that the font size of the heading message is bigger than that of the normal message
6. Send a normal single backtick code block e.g. `Hello` and observe its font size
7. Send a heading single backtick code block eg: **# `Hello`**
8. Verify that the font size of the heading message is bigger than that of the normal message

- [x] Verify that no errors appear in the JS console

### Offline tests
Not connectivity related.

### QA Steps

1. Open the app
2. Open any report
3. Send a normal message and observe its font size
4. Send a heading message (add # in beginning)
5. Verify that the font size of the heading message is bigger than that of the normal message
6. Send a normal single backtick code block e.g. `Hello` and observe its font size
7. Send a heading single backtick code block eg: **# `Hello`**
8. Verify that the font size of the heading message is bigger than that of the normal message

- [x] Verify that no errors appear in the JS console

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos

<details>
<summary>Android: Native</summary>

![android-Native](https://github.com/Expensify/App/assets/66640310/2981f0c1-57ba-4b11-91f2-93a564012872)


</details>


<details>
<summary>Android: mWeb Chrome</summary>

![android-mWeb](https://github.com/Expensify/App/assets/66640310/7af8841d-5ced-4333-bfab-5e01112bb333)


</details>


<details>
<summary>iOS: Native</summary>

![ios-native](https://github.com/Expensify/App/assets/66640310/0b6146be-e8ee-4055-a4c5-9f2cca2c46af)

</details>


<details>
<summary>iOS: mWeb Safari</summary>

![ios-mWeb](https://github.com/Expensify/App/assets/66640310/4f1df92a-c7f2-4033-befc-57859546032d)


</details>
<details>
<summary>MacOS: Chrome / Safari</summary>


https://github.com/Expensify/App/assets/66640310/34b2635c-5b59-4c50-ae43-b618eda32e19


</details>


<details>
<summary>MacOS: Desktop</summary>

![MacOS-Desktop](https://github.com/Expensify/App/assets/66640310/2e206cd4-e64b-4b36-ac94-e44a5760175a)



</details>

yakupafsin committed:
Create isInsideH1 function in htmlEngineUtils.js

Create isInsideH1 function to check if there is an ancestor node with tag name 'h1'

Signed-off-by: Yakup Afsin <yakupafsin422@gmail.com>

yakupafsin committed:
Override font size of code block for h1 tag

Check if the 'code' element is inside an 'h1' tag, and set a specific font size if true otherwise, use the font size defined in `textStyle`

Signed-off-by: Yakup Afsin <yakupafsin422@gmail.com>

yakupafsin commented:
I have read the CLA Document and I hereby sign the CLA

s77rt commented on src/components/HTMLEngineProvider/HTMLRenderers/CodeRenderer.js:
```diff
@@ -16,7 +17,13 @@ function CodeRenderer(props) {
+    const fontSize = HTMLEngineUtils.isInsideH1(props.tnode) ? 15 : textStyle.fontSize;
+
     const textStyleOverride = {
+        // Override the font family and font size for the rendered HTML element.
```
Remove unnecessary comment. Comments should explain the why not the what.

s77rt commented on src/components/HTMLEngineProvider/HTMLRenderers/CodeRenderer.js:
```diff
@@ -16,7 +17,13 @@ function CodeRenderer(props) {
         fontWeight: textStyle.fontWeight,
     });
 
+    // Check if the current element is inside an h1 tag, and set a specific font size if true,
+    // otherwise, use the font size defined in `textStyle`.
```
Remove comments

s77rt commented on src/components/HTMLEngineProvider/htmlEngineUtils.js:
```diff
@@ -44,4 +44,21 @@ function isInsideComment(tnode) {
+ * @returns {Boolean}
+ */
+function isInsideH1(tnode) {
+    let currentNode = tnode;
```
```suggestion
    let currentNode = tnode.parent;
```

We should skip the current node otherwise this may return a false positive e.g. when the tnode is actually the h1. But it's not insider another h1

s77rt replied:
@yakupafsin This is not resolved yet

s77rt commented on src/components/HTMLEngineProvider/HTMLRenderers/CodeRenderer.js:
```diff
@@ -16,7 +17,13 @@ function CodeRenderer(props) {
 
+    // Check if the current element is inside an h1 tag, and set a specific font size if true,
+    // otherwise, use the font size defined in `textStyle`.
+    const fontSize = HTMLEngineUtils.isInsideH1(props.tnode) ? 15 : textStyle.fontSize;
```
Styles should not be inlined. Write a style function that takes isInsideH1 as a parameter and returns the font size.

yakupafsin committed:
Unnecessary comment has been romeved

Signed-off-by: Yakup Afsin <yakupafsin422@gmail.com>

yakupafsin committed:
Creating style function for code tag fontSize

Signed-off-by: Yakup Afsin <yakupafsin422@gmail.com>

yakupafsin committed:
Applying getCodeFontSize function into CodeRenderer.js

Signed-off-by: Yakup Afsin <yakupafsin422@gmail.com>

yakupafsin committed:
Modifying isInsideH1 function in htmlEngineUtils.js

Signed-off-by: Yakup Afsin <yakupafsin422@gmail.com>

s77rt commented:
Please update the test steps to the following:

1. Open the app
2. Open any report
3. Send a normal message and observe its font size
4. Send a heading message (add # in beginning)
5. Verify that the font size of the heading message is bigger than that of the normal message
7. Send a normal single backtick code block e.g. `Hello` and observe its font size
8. Send a heading single backtick code block eg: # `Hello`
5. Verify that the font size of the heading message is bigger than that of the normal message

s77rt commented on src/components/HTMLEngineProvider/HTMLRenderers/CodeRenderer.js:
```diff
@@ -16,7 +17,16 @@ function CodeRenderer(props) {
 
+    // Determine the font size for the code based on whether it's inside an H1 element.
+    const fontSize = StyleUtils.getCodeFontSize(
+        HTMLEngineUtils.isInsideH1(
+            props.tnode
+        ),
```
For readability, let's define this first as a variable then pass it to the function

s77rt commented on src/styles/StyleUtils.ts:
```diff
@@ -606,6 +606,16 @@ function getFontFamilyMonospace({fontStyle, fontWeight}: TextStyle): string {
+/**
+ * Returns the font size for the HTML code tag renderer.
+ */
+function getCodeFontSize(isInsideH1:Boolean, textStyle :TextStyle) {
+    if (isInsideH1) {
+      return 15;
+    } else {
+      return textStyle.fontSize;
+    }
+  }
```
It may be better to remove the fontSize from the textStyle and just use 13 directly.

In `styles.ts` we should state that the font size is calculated by this function

yakupafsin committed:
define isInsideH1 as a boolean variable

Signed-off-by: Yakup Afsin <yakupafsin422@gmail.com>

yakupafsin committed:
Modifying `getCodeFontSize` function

Signed-off-by: Yakup Afsin <yakupafsin422@gmail.com>

yakupafsin committed:
Mentioned fontSize controlled in another function

Mentioned that fontSize controlled by getCodeFontSize function in `StyleUtils.js`

Signed-off-by: Yakup Afsin <yakupafsin422@gmail.com>

yakupafsin committed:
Run Prettier for code formatting

Signed-off-by: Yakup Afsin <yakupafsin422@gmail.com>

s77rt commented on src/components/HTMLEngineProvider/htmlEngineUtils.js:
```diff
@@ -44,4 +44,21 @@ function isInsideComment(tnode) {
+ * @returns {Boolean}
+ */
+function isInsideH1(tnode) {
+    let currentNode = tnode;
+    while (currentNode.parent) {
```
```suggestion
    let currentNode = tnode.parent;
    while (currentNode) {
```

s77rt commented on src/styles/StyleUtils.ts:
```diff
@@ -606,6 +606,16 @@ function getFontFamilyMonospace({fontStyle, fontWeight}: TextStyle): string {
+ * Returns the font size for the HTML code tag renderer.
+ */
+function getCodeFontSize(isInsideH1: Boolean) {
+    if (isInsideH1) {
+        return 15;
+    } else {
+        return 13;
+    }
```
```suggestion
    return isInsideH1 ? 15 : 13;
```

s77rt commented:
Please update the tests and the qa tests as mentioned above. Also please complete the screenshots for the rest of the platforms

s77rt commented:
## Reviewer Checklist

- [X] I have verified the author checklist is complete (all boxes are checked off).
- [X] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [X] I verified the steps for local testing are in the `Tests` section
    - [X] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [X] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [X] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [X] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [X] I verified tests pass on **all platforms** & I tested again on:
    - [X] Android: Native
    - [X] Android: mWeb Chrome
    - [X] iOS: Native
    - [X] iOS: mWeb Safari
    - [X] MacOS: Chrome / Safari
    - [X] MacOS: Desktop
- [X] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [X] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [X] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [X] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [X] I verified that comments were added to code that is not self explanatory
    - [X] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [X] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [X] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [X] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [X] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [X] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [X] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [X] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [X] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [X] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [X] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [X] If a new component is created I verified that:
    - [X] A similar component doesn't exist in the codebase
    - [X] All props are defined accurately and each prop has a `/** comment above it */`
    - [X] The file is named correctly
    - [X] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [X] The only data being stored in the state is data necessary for rendering and nothing else
    - [X] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [X] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [X] All JSX used for rendering exists in the render method
    - [X] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [X] If any new file was added I verified that:
    - [X] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [X] If a new CSS style is added I verified that:
    - [X] A similar style doesn't already exist
    - [X] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [X] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [X] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [X] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [X] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [X] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [X] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

<!-- add screenshots or videos here -->
<img width="489" alt="android" src="https://github.com/Expensify/App/assets/16493223/aea2a1cf-62f2-4e9a-a8f0-818071aa2f16">

</details>

<details>
<summary>Android: mWeb Chrome</summary>

<!-- add screenshots or videos here -->
<img width="489" alt="mweb-chrome" src="https://github.com/Expensify/App/assets/16493223/44bef25c-373f-482d-83eb-31686498ee08">

</details>

<details>
<summary>iOS: Native</summary>

<!-- add screenshots or videos here -->
<img width="574" alt="ios" src="https://github.com/Expensify/App/assets/16493223/eaed8b37-d38e-438b-a1b7-14d001036607">

</details>

<details>
<summary>iOS: mWeb Safari</summary>

<!-- add screenshots or videos here -->
<img width="574" alt="mweb-safari" src="https://github.com/Expensify/App/assets/16493223/9a9dd270-8ecc-4daf-9b3c-5117fa2d1cfe">

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

<!-- add screenshots or videos here -->
<img width="1840" alt="web" src="https://github.com/Expensify/App/assets/16493223/e27df893-2760-4195-8405-f1425a38377b">

</details>

<details>
<summary>MacOS: Desktop</summary>

<!-- add screenshots or videos here -->
<img width="1312" alt="desktop" src="https://github.com/Expensify/App/assets/16493223/19264b22-4209-424f-a915-992f0450aeaf">

</details>

s77rt commented:
@yakupafsin Thank you!

PS: Next time please tag me after making such changes because there is no notification for those actions (change title, change description, etc.)

yakupafsin commented:
@s77rt sorry, I've totally forgotten about it. Thank you for your time. And sorry for the hassle :). It was my first PR with Expensify. I'll try to do it more professionally next time.

roryabraham commented:
@yakupafsin I will not be able to merge this PR since it contains unsigned commits. Please [retroactively sign all the commits for this PR](https://webdevstudios.com/2020/05/26/retroactively-sign-git-commits/)

roryabraham commented:
You've also got lint errors it seems

yakupafsin commented:
@roryabraham because last 2 commits are co-authered woth @s77rt I guess that's why it looks unsigned. We need to undo that 2 commit and re-commit them.

yakupafsin committed:
Update htmlEngineUtils.js

Signed-off-by: Yakup Afsin <yakupafsin422@gmail.com>

yakupafsin committed:
Update StyleUtils.ts

yakupafsin commented:
@roryabraham I've removed that 2 commits and pushed 2 new commits. I guess all will be signed now.

yakupafsin commented:
@roryabraham I've double-checked but couldn't find any lint errors.

roryabraham commented:
Commits are still unsigned:

<img width="857" alt="image" src="https://github.com/Expensify/App/assets/47436092/50e2aca4-8d94-4393-99f2-7434cd22a8e1">

In order to be able to merge code in this repo, every commit must say `Verified` next to the commit hash.

yakupafsin commented:
Hi @roryabraham, I've signed all the commits. Because I've changed my device I forgot change my gitconfig file. Sorry about that. But I've signed them now.

roryabraham commented:
@yakupafsin looks like we've got some lint errors here: https://github.com/Expensify/App/actions/runs/6792085078/job/18467959217?pr=30876

Also, NAB but if you don't mind I think we could generalize `HTMLEngineUtils.isInsideH1` to be something like `isChildOfNode`:

```js
/**
 * Check if there is an ancestor node for which the predicate returns true.
 *
 * @param {TNode} tnode
 * @param {Function} predicate
 * @returns {Boolean}
 */
function isChildOfNode(tnode, predicate) {
    let currentNode = tnode.parent;
    while (currentNode) {
        if (predicate(currentNode)) {
            return true;
        }
        currentNode = currentNode.parent;
    }
    return false;
}

/**
 * Check if node is a child of a comment tag.
 *
 * @param {TNode} tnode
 * @returns {Boolean}
 */
function isChildOfComment(tnode) {
    return isChildOfNode(tnode, (node) => isCommentTag(node.domNode.name));
}

/**
 * Check if node is a child of an h1 tag.
 *
 * @param {TNode} tnode
 * @returns {Boolean}
 */
function isChildOfH1(tnode) {
    return isChildOfNode(tnode, (node) => node.domNode.name.toLowerCase() === 'h1');
}
```

yakupafsin commented:
@roryabraham I was planning to do something like this, but I decided not to touch someone else's code 😅. But yes, that looks better. Let me change it.

yakupafsin committed:
Optimize with isChildOfNode function

yakupafsin committed:
Resolve lint error

yakupafsin commented:
Hi @roryabraham, I've found the lint error. And also applied the changes on htmlEngineUtils.js

roryabraham commented:
@yakupafsin please run `npm run lint` to find and address the lint errors.

yakupafsin committed:
Using boolean instead Boolean

yakupafsin commented:
@roryabraham now it's ready to merge I guess :)

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

tgolen commented on src/components/HTMLEngineProvider/htmlEngineUtils.js:
```diff
@@ -28,20 +28,41 @@ function isCommentTag(tagName) {
+ * @returns {Boolean}
+ */
+function isChildOfH1(tnode) {
+    return isChildOfNode(tnode, (node) => node.domNode.name.toLowerCase() === 'h1');
```
FYI, this is causing staging to crash and I cannot load the page at all.

![image](https://github.com/Expensify/App/assets/1228807/2d829545-bcc6-4e1d-ba06-794f81e60744)

s77rt replied:
@yakupafsin Can you please raise a quick PR fixing this? I think we may not need the `toLowerCase` afterall. The node name for html elements is expected to always be uppercase [^1].

[^1]: https://developer.mozilla.org/en-US/docs/Web/API/Node/nodeName

roryabraham replied:
Sorry for the disruption – this has been fixed: https://github.com/Expensify/App/pull/31079

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/6826733490) to production by https://github.com/puneetlath in version: 1.3.97-7 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/6826740157) to staging by https://github.com/roryabraham in version: 1.3.98-0 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/6869834824) to production by https://github.com/roryabraham in version: 1.3.98-5 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅
